test/unit/org/apache/cassandra/index/sai/cql/StorageAttachedIndexDDLTest.java

```diff
@@ -278,43 +326,757 @@ public void shouldFailCreationOnMultipleColumns()
+                     expectedLiteralState,
+                     verifyChecksum(literalIndexContext));
+
+        if (rebuild)
+        {
+            rebuildIndexes(literalIndexContext.getIndexName());
+        }
+        else
+        {
+            // Reload all SSTable indexes to manifest the corruption:
+            reloadSSTableIndex();
+
+            // Verify the index cannot be read:
+            verifySSTableIndexes(literalIndexContext.getIndexName(), Version.LATEST.onDiskFormat().perSSTableComponents().contains(component) ? 0 : 1, 0);
+
+            try
+            {
+                // If the corruption is that a file is missing entirely, the index won't be marked non-queryable...
+                rows = executeNet("SELECT id1 FROM %s WHERE v1='0'");
+                assertEquals(0, rows.all().size());
+            }
+            catch (ReadFailureException e)
+            {
+                // ...but most kind of corruption will result in the index being non-queryable.
+            }
+
+            // Simulate the index repair that would occur on restart:
+            runInitializationTask();
+        }
+
+        // verify indexes are recovered
+        verifySSTableIndexes(literalIndexContext.getIndexName(), 1);
+        verifyIndexFiles(literalIndexContext, 1, 1, 1);
+
+        rows = executeNet("SELECT id1 FROM %s WHERE v1='0'");
+        assertEquals(rowCount, rows.all().size());
+    }
+
+    @Test
+    public void verifyCleanupFailedPerIndexFiles() throws Throwable
+    {
+        createTable(CREATE_TABLE_TEMPLATE);
+        disableCompaction(KEYSPACE);
+
+        execute("INSERT INTO %s (id1, v1) VALUES ('0', '0');");
+        flush();
+        execute("INSERT INTO %s (id1, v1) VALUES ('1', '0');");
+        flush();
+
+        // Inject failure
+        Injections.inject(failPerIndexMetaCompletion);
+        failPerIndexMetaCompletion.enable();
+
+        try
+        {
+            // Create a new index, which will actuate a build compaction and fail, but leave the node running...
+            IndexContext literalIndexContext = createIndexContext(createIndex(String.format(CREATE_INDEX_TEMPLATE, "v1")), UTF8Type.instance);
+            // two index builders running in different compaction threads because of parallelised index initial build
+            waitForAssert(() -> assertEquals(2, indexBuildCounter.get()));
+            waitForCompactionsFinished();
+
+            // Only token/primary key files for the first SSTable in the compaction task should exist, while column-specific files are blown away:
+            verifyIndexFiles(literalIndexContext, 2, 0, 0);
+
+            assertEquals("Segment memory limiter should revert to zero.", 0L, getSegmentBufferUsedBytes());
+            assertEquals("There should be no segment builders in progress.", 0L, getColumnIndexBuildsInProgress());
```

Comment:
After this line, I added...

```
waitForIndexQueryable();
executeNet("SELECT id1 FROM %s WHERE v1='0'");
```
...and the test still passes. Do we actually want the index to be queryable when there's data in the backing table that's clearly missing from the index? What am I missing?

(same issue in `verifyCleanupFailedTokenPrimaryKeyFiles`, btw)

Reply:
This definitely looks to be a bug and a lack of testing. I'm confirming this and will fix it.

Reply:
This was happening because the `StorageAttachedIndexWriter` was not propagating errors when it was called as part of an index build operation. I have added a flag to its constructor to allow errors to be propagated if this is the case but not if it part of a flush or compaction operation. This has fixed the problem and I have added an `assertFalse(isIndexQueryable)` to the tests to ensure that this is the case.